library/alloc/src/rc.rs

```diff
@@ -2512,14 +2512,21 @@ trait RcInnerPtr {
     fn inc_strong(&self) {
         let strong = self.strong();
 
+        // We insert an `assume` here to hint LLVM at an otherwise
+        // missed optimization.
+        // SAFETY: The reference count will never be zero when this is
+        // called.
+        unsafe { core::intrinsics::assume(strong != 0); }
+
+        let strong = strong.wrapping_add(1);
+        self.strong_ref().set(strong);
+
         // We want to abort on overflow instead of dropping the value.
-        // The reference count will never be zero when this is called;
-        // nevertheless, we insert an abort here to hint LLVM at
-        // an otherwise missed optimization.
-        if strong == 0 || strong == usize::MAX {
+        // Checking after the store instead of before allows for
+        // slightly better code generation.
+        if core::intrinsics::unlikely(strong == 0) {
```

Comment:
Sorry if I'm mistaken, but shouldn't this be `unlikely(strong == usize::MAX)`?

Reply:
Same below.

Reply:
No, because `strong.wrapping_add(1)` is executed before this if-statement. The original check did this before the increment; in other words it was checking if overflow *will* occur. But this new code checks after the increment, or in other words it's checking if overflow *did* occur.

Reply:
Ah, right. Imo the order should either be changed or a comment be added, because at first glance it looks wrong.

Reply:
There is a comment, but perhaps it wasn't clear enough. I've now changed it from "Checking after the store instead of before allows for slightly better code generation" to "Checking _for overflow_ after the store instead of before allows for slightly better code generation" (new addition in italics).